.github/workflows/publish.yml

```diff
@@ -0,0 +1,49 @@
+# This workflow is triggered two ways:
+#
+# 1. When a tag is created, the workflow will upload the package to
+#    test.pypi.org.
+# 2. When a release is made, the workflow will upload the package to pypi.org.
+#
+# It is done this way until PyPI has draft reviews, to allow for a two-stage
+# upload with a chance for manual intervention before the final publication.
+name: Upload package
+
+on:
+  release:
+    types: [created]
+  push:
+    tags:
+      - '*'
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python
+      uses: actions/setup-python@v1
+      with:
+        python-version: '3.x'
+    - name: Install dependencies
```

Comment:
You may want to also inject a cache task for pip.

Reply:
Hm, do you have a good example of this in use? I found [this](https://github.com/actions/cache/blob/master/examples.md#python---pip), but it's not clear to me whether that is supposed to go before the "install pip" stage or after it. I assume after?

In any case, I'm not sure it's necessary for this job – I suspect there will be a *lot* of cache misses; I expect this job to run between 6 and 20 times per year, so just in the time it took me to consider and write up this comment I've probably blown the total time savings from a cache for the next decade :stuck_out_tongue_closed_eyes: (though I'll be glad to know about it for my other stuff which *does* run many times per day).

Reply:
It should be before pip install because it fetches the caches and puts them on disk. It also has a post-workflow step that saves things later. But yeah, technically you may ignore it if you want to.

Reply:
Btw, following your article about testing projects as installed I'm working on a state-of-the-art workflow that builds dists, then tests them (pip installing from sdist/wheels) and only then does the upload of what's tested. I'll be ready to show it off soon, hopefully.

Reply:
> It should be before pip install because it fetches the caches and puts them on disk. It also has a post-workflow step that saves things later. But yeah, technically you may ignore it if you want to.

Interesting, though, because this example requires `pip`, and a recent version at that: https://github.com/actions/cache/blob/master/examples.md#using-pip-to-get-cache-location

Are they suggesting an "install pip" stage, then a "pip cache" stage, then an "install dependencies" stage?



> Btw, following your article about testing projects as installed I'm working on a state-of-the-art workflow that builds dists, then tests them (pip installing from sdist/wheels) and only then does the upload of what's tested. I'll be ready to show it off soon, hopefully.

Yes please! I tried to build some half-assed version of this for `dateutil` 5 years ago to be manually triggered but eventually I gave up. I doubt that it will catch too many problems if you're already using a "test-as-installed" workflow, but it would certainly help my peace of mind with automated deploys.

Reply:
> Interesting, though, because this example requires `pip`, and a recent version at that: [actions/cache:examples.md@`master`#using-pip-to-get-cache-location](https://github.com/actions/cache/blob/master/examples.md#using-pip-to-get-cache-location) [](/actions/cache/blob/HEAD@{2020-05-06T12:56:17Z}/examples.md#using-pip-to-get-cache-location)

That's just for dynamically getting the cache location rather than hardcoding. But the action itself grabs the cache archive and unpacks it right where you stick the step so any next step would have access to files on disk. And if pip modifies that cache along the way, when all steps of the job finish, it'll compile a new archive and upload it.




> Yes please! I tried to build some half-assed version of this for `dateutil` 5 years ago to be manually triggered but eventually I gave up. I doubt that it will catch too many problems if you're already using a "test-as-installed" workflow, but it would certainly help my peace of mind with automated deploys.

Well, it's here but some bits are still in progress (like testing sdist on ubuntu and uploading is also disabled there) https://github.com/ansible/pylibssh/actions?query=workflow%3A%22%F0%9F%8F%97+%F0%9F%93%A6+%26+test+%26+publish%22